src/Octoshift/Services/AdoPipelineTriggerService.cs

```diff
@@ -0,0 +1,434 @@
+/// </summary>
+public class AdoPipelineTriggerService
+{
+    private readonly AdoApi _adoApi;
+    private readonly OctoLogger _log;
+    private readonly string _adoBaseUrl;
+
+    public AdoPipelineTriggerService(AdoApi adoApi, OctoLogger log, string adoBaseUrl)
+    {
+        _adoApi = adoApi ?? throw new ArgumentNullException(nameof(adoApi));
+        _log = log ?? throw new ArgumentNullException(nameof(log));
+        _adoBaseUrl = adoBaseUrl?.TrimEnd('/');
+    }
+
+    /// <summary>
+    /// Changes a pipeline's repository configuration from ADO to GitHub, applying
+    /// trigger configuration based on branch policy requirements and existing settings.
+    /// </summary>
+    public virtual async Task RewirePipelineToGitHub(
+        string adoOrg,
+        string teamProject,
+        int pipelineId,
+        string defaultBranch,
+        string clean,
+        string checkoutSubmodules,
+        string githubOrg,
+        string githubRepo,
+        string connectedServiceId,
+        JToken originalTriggers = null,
+        string targetApiUrl = null)
+    {
+        var url = $"{_adoBaseUrl}/{adoOrg.EscapeDataString()}/{teamProject.EscapeDataString()}/_apis/build/definitions/{pipelineId}?api-version=6.0";
+
+        var response = await _adoApi.GetAsync(url);
+        var data = JObject.Parse(response);
+
+        var newRepo = CreateGitHubRepositoryConfiguration(githubOrg, githubRepo, defaultBranch, clean, checkoutSubmodules, connectedServiceId, targetApiUrl);
+        var currentRepoName = data["repository"]?["name"]?.ToString();
+        var isPipelineRequiredByBranchPolicy = await IsPipelineRequiredByBranchPolicy(adoOrg, teamProject, currentRepoName, pipelineId);
+
+        LogBranchPolicyCheckResults(pipelineId, isPipelineRequiredByBranchPolicy);
+
+        var payload = BuildPipelinePayload(data, newRepo, originalTriggers, isPipelineRequiredByBranchPolicy);
+
+        await _adoApi.PutAsync(url, payload.ToObject(typeof(object)));
+    }
+
+    /// <summary>
+    /// Analyzes branch policies to determine if a pipeline is required for branch protection.
+    /// </summary>
+    public async Task<bool> IsPipelineRequiredByBranchPolicy(string adoOrg, string teamProject, string repoName, int pipelineId)
+    {
+        if (string.IsNullOrEmpty(repoName))
+        {
+            _log.LogWarning($"Branch policy check skipped for pipeline {pipelineId} - repository name not available. Pipeline trigger configuration may not preserve branch policy requirements.");
+            return false;
+        }
+
+        try
+        {
+            // Get repository information first
+            var repoUrl = $"{_adoBaseUrl}/{adoOrg.EscapeDataString()}/{teamProject.EscapeDataString()}/_apis/git/repositories/{repoName.EscapeDataString()}?api-version=6.0";
+            var repoResponse = await _adoApi.GetAsync(repoUrl);
+            var repoData = JObject.Parse(repoResponse);
+            var repositoryId = repoData["id"]?.ToString();
```

Comment:
How expensive is this to invoke? It'll be called once per pipeline, even if it can be reused. It seems likely that this would be better to be cached.